Important PR: Enhancement - restic snapshot filters based on time

wplapper opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR introduces time based filtering along the lines of https://github.com/restic/restic/issues/5193.
<!--
Describe the changes and their purpose here, as detailed as needed.
-->
All commands which use a ``SnapshotFiler`` gain three additional options: ``--lower-time-limit`` and ``--upper-time-limit`` and ``--relative-to``. These options limit the time range the user is interested in. Since #5193 suggests three methods of defining a time reference, one of them is a relative time, something like "two and a half months ago", a reference time is needed. That is what ``--relative-to`` will provide.

A new concept ``DurationTime`` is introduced, which can be one of three things: a ``time.Time``, a ``restic.Duration`` related to a reference Time (``RelativeTo``) or a snapshot ID, including 'latest'. All these will eventually be converted to a ``time.Time``, so the ``matches()`` function can do time based filtering ``--lower-time-limit <= sn.Time <= --upper-time-limit``. 

The main part of the new code is located in ``internal/restic/snapshot_find.go``, which is handling the interface to the command line (Set()), which deals with interpretation of the user's input by using regexp() for the definition of a ``restic.Duration`` and the definition of a date with an optional time. The strings ``now`` and ``latest`` are interpreted in the correct context. 

The methods ``Empty()`` and ``matches()`` are extended so they can deal with time based filters.

If a time based filter is detected which references a snapshot or ``latest``, the function ``setTimeFilters()`` is called to convert these snapshot IDs to their equivalent time values. Finally ``setTimes()`` is called to check the sanity of the filter values and converts a ``restic.Duration`` to its equivalent time.Time, based on the value of ``--relative-to`` option which is either set explicitly or implicitly by using the snapshot ID of ``latest``. All of the above happens in function ``FindAll()``.

The method String() for a ``DurationTime`` reads the state of the ``DurationTime`` element and creates a string representation which matches one of a ``snapshot ID``, a ``restic.Duration`` or a ``time.Time``.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
https://github.com/restic/restic/issues/2078 // More options for "restic tag"
https://github.com/restic/restic/issues/5136 // Ability to filter snapshots by date
https://github.com/restic/restic/issues/5193 // Allow forget to relate policy to current time and not just latest snapshot
https://github.com/restic/restic/pull/5209   // Add snapshot filtering --lower-time-limit and --upper-time-limit
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Alternative names for time based filters could be ``--filter-before`` and ``--filter-after``; ``rustic`` is using these names.

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

~~- [ ] I have added tests for all code changes.~~
  see separate PR https://github.com/restic/restic/pull/21936 for tests.
- [x] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [X] I'm done! This pull request is ready for review.

wplapper committed:
restic time based filters

This refers to all commands which use a data.SnapshotFilter for filtering.

cmd/restic/find.go:
add filter definitions `--older-than`, `--newer-than`and `--relative-to`

internal/data/duration.go:
add definition of DuarationTime, a concept which holds one of
- a snapshot ID including 'latest'
- a data.Duration time difference, e.g. '11m15d12h'
- an absolute date or time in the format 'yyyy-mm-dd' or 'yyyy-mm-dd HH:MM:SS' or the string 'now'

add the functions Set(), String(), Type() as standard interfaces
and Empty(), AddOffset() and GetTime() to interface with these new type.

internal/data/snapshot_find.go:
extends the functions Empty(), matches().

Adds code to FindAll(), so that DurationTime is converted to an absolute time, dependent on the
initial setting from the command line.

This adds functions
- buildSnapTimes() which is called from FindAll() to access DurationTime elements
- setTimeFilters() for evaluation of snapID based time elements,
- setTimes() to convert all Duration elements and do some error checking

internal/data/duration_test.go and cmd/restic/cmd_snapshots_integration_test.go:
add tests

changelog/unreleased/issue-5193:
describe the enhancement for time based filters

doc/045_working_with_repos.rst:
describes time based filters with examples

windows test failures

get rid of these pesky windows tests - temporarily

restic SnapshotFilter - restore windows tests

internal/restorer/restorer_windows_test.go - let the windows tests be run again!

wplapper committed:
restic SnapshotFiler --older-than & --newer-than

Complete rework of the PR, since too many things have changed

cmd/restic/find.go:
setting up for SnapshotFilter

internal/data/duration.go:
definition of SnapshotFilter

internal/data/snapshot_find.go:
working internally for the SnapshotFilter

cmd/restic/cmd_snapshots_integration_test.go:
internal/data/duration_test.go:
internal/data/snapshot_find_test.go:
tests for the snapshot filter

internal/backend/local/local_windows.go:
gofmt created this change - whitespace

wplapper committed:
restic time based filters - integration tests

during rebase I had to remove the additional tests from cmd_snapshots_integration_test.go.
They are re-introduced now.

restic time based filters - remove debugging code

removed all references to DurationTime.name

wplapper committed:
time based filters - add debug.Log()

Added 5 debug.Log() statements to show the active filter components
of a snapshot filter.

wplapper committed:
time based filters - more debug.Log()

added one more debug.Log() statement in setTimes()

wplapper committed:
Duration and DurationTime are now separate files

Duration and DurationTime are now moved into theit own files.
Therefore the tests also had to be split into two different files.
The logic is still the same.

time base snapshot filters - minor corrections

internal/data/duration_time.go: stricter checking for logic errors

internal/data/snapshot_find.go: rewording of some comments
changed to 'if err := ...; err != nil' clause

internal/data/snapshot_find_test.go:
removed duplicate tests for data.DurationTime

wplapper commented:
Some tests still fail - will try to re-run it again later.

wplapper commented:
temporary close to trigger rerun of test jobs.

rawtaz commented:
@wplapper You don't have to write a comment for every action you perform to manage the code on this branch.

E.g. things like pushing commits is common practice, even if you force-push sometimes. And closing and reopening a PR is presumed to be something you do intentionally or unintentionally, but what matters is if you keep it open at the end of the day, since that means you still want it to be active.

My point here is that all these comments are just creating noise, so maybe try to focus on comments that add to the discussion instead – we have faith in that you're taking actions that are for the benefit of the PR, there's no need to prove that :)

wplapper commented:
> @wplapper You don't have to write a comment for every action you perform to manage the code on this branch.

Understood. I removed all "noisy comments".

wplapper committed:
SnaphotFiler based on timestamps

For the first anniversary of this PR I decided to do some spring cleaning
renaming the confusing --older-than and --newer-than to --lower-time-limit and
--upper-time-limit. As a consequence, the following files were modified

* cmd/restic/find.go:
modified the definition and removed comment referencing non-existent
restic.FindFilteredSnapshots()

* internal/data/snapshot_find.go:
renamed the variables to the new naming conventions
added a new function FormatTimeRange() which is used for error messages and for
debug output
condensed debugging output

* internal/data/duration_time.go
made sure that returned time.Time are Local()

* cmd/restic/cmd_snapshots_integration_test.go
renamed local variables to the new conventions for the tests and
added two tests for conrer cases

wplapper committed:
SnapshotFilter PR 1st anniversary

* cmd/restic/cmd_snapshots_integration_test.go
added error check for setting test variables

* changelog/unreleased/issue-5193
changed announcement file to the new naming conventions

wplapper committed:
SnapshotFilter 1st anniversary

* changelog/unreleased/issue-5193:
shorten announcement details and fix first line

* internal/data/duration_time.go:
fix linter error

wplapper committed:
SnapshotFiler time based documentation changes

* doc/045_working_with_repos.rst adopted the new naming convention

wplapper committed:
Snapshotfilter for time - docs

fix typo

wplapper committed:
SnapshotFilter - by date/time

added an alias of 'newer-than' = 'lower-time-limit' and
'older-than' = 'upper-time-limit'.
Sometimes it can be a bit confusing what these boudary options mean.

wplapper committed:
SnapshotFilters based on time - docs

correct 2 spelling mistakes

filters based on time - fix typo

omitted "}" while rebasing to the latest master

find.go - merged incorrect comment

soemhow the incorrect comment was merged into find.go regarding
initSingleSnapshotFilter()

wplapper committed:
snapshot tests - suppress double snapshot listing error

wplapper committed:
snapshot filters based on time - removed tests

removed tests to make package smaller

wplapper commented:
I split off cmd/restic/cmd_snapshots_integration_test.go and internal/data/duration_time_test.go to make the review package smaller. The rest will follow in a separate PR.

wplapper commented:
The tests are in PR https://github.com/restic/restic/pull/21936.
